tools/crm_rule.c

```diff
@@ -44,7 +44,19 @@ static struct crm_option long_options[] = {
     {"-spacer-",   required_argument, NULL, '-', "\n\nThis tool is currently experimental.",
      pcmk_option_paragraph},
     {"-spacer-",   required_argument, NULL, '-', "The interface, behavior, and output may "
-                                                 "change with any version of pacemaker.", pcmk_option_paragraph},
+                                                 "change with any version of pacemaker.",
+     pcmk_option_paragraph},
+    {"-spacer-",   required_argument, NULL, '-', "Important: avoid making conclusion based on"
+                                                 " this tool when it is not of the same version as"
+                                                 " an overall pacemaker deployment."
+                                                 "\nThe exact date/time parsing behaviour may get altered"
+                                                 " over time, so only the above requirement guarantees"
+                                                 " reliable outcomes related to particular deployment"
+                                                 " (i.e. without a generic validity for older/newer"
+                                                 " versions of pacemaker, possibly -- depending on the"
+                                                 " permanency of the arising decisions -- mandating"
+                                                 " any such to be re-evaluated anew upon an update).",
+     pcmk_option_paragraph},
```

Comment:
Please remove this note. It engenders more confusion than clarity, and man pages / help output is not the right place to document release notes.

Reply:
Agree it sounds maybe too harsh.

Do you suggest to document the expected problems arising from
real parser bugs elsewhere?

The "closure" makes it as least good-in-bad, so shall be documented
somewhere...

Reply:
Bugs are documented in the change log. We simply don't need to warn users about fixed bugs. There are at least hundreds of potential problems with mixed-version clusters due to bugs that are fixed in the newer version. The documentation already strongly recommends that mixed versions be used only during a brief rolling upgrade window to minimize the chance for problems. Also, the fact that no one's ever reported the bug suggests that it's a rare usage; the average user shouldn't be bothered with it.

Reply:
The problem is that we should educate users to use the software
correctly wrt. slight changes in behaviour and _precomuputed assumotions_
-- it's much safer to re-evaluate some decision making variables in the
game after each update than to stay totally static there.

Especially when/if the parser is to be changed completely one
day -- CIB is mostly fine, it's the state we have access to and
can _reconcile_ the situation with it.  But we don't have the
visibility into all the surrounding scripting etc. so that's why
I think it'd be wise to lead an example of very dynamic adaptations
whereever possible.

It's not limited to date/time parsing as such, it shall be a very
general guidance, actually.

Reply:
Very contrived, next to bogus example, to demonstrate what I mean:

- user has a home-grown algorithm to allocate new cron tasks (local
  only) it is given as uniformly throughtout the day as possible (for
  some reason it is critical for the overall business)

- as a last step, this algoritm consults `crm_rule` and crossmatches
  selected rule `id`s against the time currently predestined for the
  new task to be scheduled

- so far so good, now two branches

a) user upgrades pacemaker, and suddenly, the rules containing
   a timezone without space gets the offset with moved precision
   from units of hours to tens of minutes (as we have just seen)

   - only documentation may help, since it the new pacemaker not
     fitting the old state of affairs (due to a slight behaviour
     altering)

b) user upgrades pacemaker, is educated only partly so that she
   rushes to have the above cron tasks rescheduled reflecting
   new answers to "is given rule active at given point in time?",
   but alas, forgetting she can't make these decisions based
   on other version of the pacemaker in her simulated cluster
   environments using VMs on here laptop

   - the above documentation shall highlight the necessity
     of version/scope lock-step for decisions about the
     "system" to stay on the safest side

Now, how to sell it to the users, that SW perfection is unattainable
in general, and that pieces that are not totally self-contained
(without effect on the outer world) -- which pacemaker as a distributed
system can hardly be (causing so many compatibility headaches) -- and
hence some level of users' cooperation and assistance and carefulness
is needed so they can live in peace and accord with such SW...